face-detection-explainer.md

```diff
@@ -12,66 +12,129 @@
+* Face detection API does not need to return a mesh corresponding to the detected faces. Even though TensorFlow returns a 468-landmark FaceMesh and most DNNs can return something similar, mesh is not supported on any platforms presently, and for the sake of simplicity, it is excluded for now. However, in the long term it may be appropriate to extend the face detection API to be able to also return mesh-based face detection results. This is left for future work.
 
-## Using the face detection API
+## Face detection API
 
+### Metadata
 
-The API consists of two parts: first, the constraint `faceDetectionMode` to `getUserMedia()` or `applyConstrains()` is used to negotiate and enable the desired face detection mode. Often, camera drivers run already internally face detection for 3A, so enabling face detection might just make the results available to Web applications. The corresponding fields are added also to media track settings and capabilities. 
+The API consists of two parts: first, metadata which describes the faces available in video frames. The metadata could be also set by user by creating a new video frame (modifying video frame metadata in existing frames is not allowed by the specification). The metadata could be used by WebCodecs encoders to improve video encoding quality, for example by allocating more bits to face areas in frames. As the encoding algorithms are not specified in standards, also the exact way how the facial metadata is used is not specified.
 
-The second part of the API adds a new field `detectedFaces` of type sequence of `DetectedFace` into WebCodecs `VideoFrameMetadata` which provides information of the detected faces on the frame. In `DetectedFace`, the field `id` is used to track faces between frames: the same non-zero `id` in a face between different frames indicates that it is the same face. The value of zero means that tracking is not supported. `probability` is the probability that the returned face is in fact a face and not a false detection. This should be always greater than zero but less than one, since no perfect algorithm exists. If probability is not specified, user agent sets the field to zero.
+```js
+partial dictionary VideoFrameMetadata {
+  sequence<HumanFace> humanFaces;
+};
 
-The field `contour` provides an arbitrary number of points which enclose the face. Sophisticated implementations could provide a large number of points but initial implementations are expected to provide four contour points located in the corners of a rectangular region which describes the face bounding box. User agent is allowed to provide a minimum of one point, in which case the point should be located at the center of the face.
+dictionary HumanFace {
+  long              id;
+  float             probability;

```

Comment:
This is probably better expressed as a "confidence".

It's not really a probability: the face is either there or not.  This value is intended to represent how confident the API is that the face is there.

Reply:
I assume that here "probability" is the estimated probability provided by the model, based on the data it was trained on.

Confidence is typically not expressed as a single number, but rather a range.

Reply:
> I assume that here "probability" is the estimated probability provided by the model,

Naming aside, leaving the range of possible values implementation-specific may make it hard for apps to reason about the values. E.g. if one implementation's model rarely ever outputs above 0.7 for some technical purity reason, even in a well-lit face front situation, and another model reliably produces 1.0 except in corner cases, it could lead to weird web compat issues.

Are there precedents we could look at? Would it make sense to say something like `1.0` SHOULD represent the highest value the model can consistently output?

Reply:
I think confidence is the right term.
> Each confidence estimate might be seen as a probability of correctness, reflect-
ing the model’s uncertainty about a detection or the pixel mask. During inference,
we expect the estimated confidence to match the observed precision for a predic-
tion. For example, given 100 predictions with 80% confidence each, we expect 80
predictions to be correctly predicted
https://arxiv.org/pdf/2202.12785

That very paper describes how to ensure proper calibration of this confidence for object detection; I'm not sure whether we want to go to that level of specificity though.

Reply:
> leaving the range of possible values implementation-defined may make it hard for apps to reason about the values.

I agree with this 100% and think that unless the value is properly defined it shouldn't be included into the spec.

The paper that dom quotes says "confidence estimate **might be seen** as a probability of correctness".  To me this doesn't look like very rigorous or widely used definition. I'd rather go to commonly used definitions, such as [Wikipedia](https://en.wikipedia.org/wiki/Probability). I believe what we would like to have here is conditional probability, but the condition needs to be more precisely defined than what it used to be. In the latest explainer/spec draft I changed the definition as follows:

"the estimate of the conditional probability that the segmented object actually is of the type indicated by the {{Segment/type}} member on the condition that the detection has been made." (where {{Segment/type}} describes whether it is a face or some other type of object; see [issue 79](https://github.com/w3c/mediacapture-extensions/issues/79).

It is an estimate, not exact probability. As an implementor advise, the implementor could run the face detection algorithm on some test videos and of the detections that the algorithm made, check how many are correct (missed faces would not affect the conditional probability here although the algorithm quality affects that too). Other factors could also be taken into account, such as confidence/score from the algorithm, camera image quality, etc.